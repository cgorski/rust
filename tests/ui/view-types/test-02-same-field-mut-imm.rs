#![feature(view_types)]
#![allow(unused, incomplete_features)]

// TEST 2: Theorem 4 - same_field_mut_imm_conflicts
//
// THEOREM (formalization/Core_Proven.v):
// Theorem same_field_mut_imm_conflicts : forall f,
//   fields_conflict (mkFieldAccess f Mut) (mkFieldAccess f Imm) = true.
//
// MEANING: A mutable access to a field conflicts with an immutable access
//          to the same field
//
// PROOF TECHNIQUE: Same field name, one mutable â†’ always conflict
//
// TEST STRATEGY:
// Create mutable borrow of a field, then call view-typed method that
// immutably accesses that field. Should error (cannot have mutable and
// immutable borrow of same data simultaneously).

struct S {
    a: i32,
    b: i32,
}

impl S {
    // View spec: immutably accesses field 'a'
    fn read_a(&{a} self) -> i32 {
        self.a
    }
}

fn main() {
    let mut s = S { a: 0, b: 0 };

    // Explicitly borrow field 'a' mutably
    let a_ref = &mut s.a;

    // Try to call method that immutably accesses 'a'
    // Theorem 4 guarantees this conflicts (mut + imm on same field)
    s.read_a(); //~ ERROR cannot borrow

    // Keep the first borrow active
    *a_ref = 10;
}
