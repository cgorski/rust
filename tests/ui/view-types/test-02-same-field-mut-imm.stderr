error[E0502]: cannot borrow `s.a` as immutable because it is also borrowed as mutable
  --> $DIR/test-02-same-field-mut-imm.rs:40:5
   |
LL |     let a_ref = &mut s.a;
   |                 -------- mutable borrow occurs here
...
LL |     s.read_a();
   |     ^ immutable borrow occurs here
...
LL |     *a_ref = 10;
   |     ----------- mutable borrow later used here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0502`.
